src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -68,6 +105,42 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
     /* Initialize other global parameters */
     MPIDI_XPMEM_global.sys_page_sz = (size_t) sysconf(_SC_PAGESIZE);
 
+    void *coop_counter_vaddr = NULL;
+    OPA_int_t *coop_counter = NULL;
+    void *rank0_coop_counter_addr = NULL;
+    /* Only rank 0 need to actually allocate an array to store counters and
+     * other processes should attach to this array through MPIDI_XPMEM_seg_regist */
+    if (!MPIR_Process.local_rank) {
+        coop_counter =
+            (OPA_int_t *) MPL_aligned_alloc(MPIDI_XPMEM_global.sys_page_sz,
+                                            num_local * num_local * sizeof(OPA_int_t),
+                                            MPL_MEM_OTHER);
+        MPIR_ERR_CHKANDJUMP(coop_counter == NULL, mpi_errno, MPI_ERR_OTHER, "**nomem");
+
+        for (i = 0; i < num_local * num_local; ++i)
+            OPA_store_int(&coop_counter[i], 0);
+    }
+
+    MPIDU_Init_shm_barrier();
+    MPIDU_Init_shm_put(&coop_counter, sizeof(OPA_int_t *));
+    MPIDU_Init_shm_barrier();
+    MPIDU_Init_shm_get(0, sizeof(void *), (void *) &rank0_coop_counter_addr);
+
+    /* Attach to local root coop_counter array */
+    if (MPIR_Process.local_rank) {
+        mpi_errno =
+            MPIDI_XPMEM_init_seg_regist(0, num_local * num_local * sizeof(OPA_int_t),
+                                        rank0_coop_counter_addr, (void **) &coop_counter,
+                                        &coop_counter_vaddr);
+        if (mpi_errno != MPI_SUCCESS)
```

Comment:
We need to distinguish public routines and internal routines. It is very confusing to use the same prefix for both. Ideally, we do not need name space (e.g. `MPIDI_XPMEM_`) for any internal routine. When we define internal routines in a header file, however, we cannot eliminate prefix otherwise it may conflict with a function using the same name in another header file. But here is a C file.

Reply:
I can add `static` at the beginning of the function, it would be used only as local function.

Reply:
`static` is necessary, but could you please also following the naming rule? It is a rule applied to all MPICH code.

Reply:
ok